_posts/2019-01-10-future-workloads.md

```diff
@@ -0,0 +1,503 @@
+
+##### Asynchronous pool
+
+These futures can usually be handled in a single thread. They are simply dispatching values so they should be very fast.
+
+We can choose to schedule these in heterogeneous workloads but this can introduce turnaround problems, which means these faster futures may be blocked by long running futures. Separating these may make sense in order to improve throughput.
+
+##### CPU bound pool
+
+When doing cpu bounded work, it is best to use a thread pool size equal to the **number of cores**. After all, we are keeping all cores busy anyway so adding extra threads should only degrade performance with the extra context switching, kernel calls and thread memory usage.
+
+The **scala.concurrent.ExecutionContext.Implicits.global** is probably the most widely used pool for this kind of workload. As mentioned before, this executor uses a fork join thread pool to manage the workloads. This type of executor is the best fit for highly asynchronous workloads but is less required when they are less asynchronous, because we don't have to deal with that much contention and stealing, which this pool tries to solve.
+
+We can explain the level of concurrency for our workloads as such:
+
+| ms        | Method        | Concurrency   |
+|-----------|---------------|---------------|
+| 0         | asynchronous  | Highly asynchronous |
+| 1         | Future(cpuBound(1)) | More asynchronous  |
+| 5         | Future(cpuBound(5)) | Less asynchronous  |
+| x       | cpuBound(x)   | Synchronous |
+
+We can also visualize it as a function of the `cpuBound` time parameter. A `Future(cpuBound(x))` will be more asynchronous the closer it is to `1 ms`.
+
+```
+ms
+|1------------------------------------------Inf|
+```
+
+#### Implicitly or explicitly
+
+We looked at the `Future.apply` method definition before. It expects an implicit `ExecutionContext`, and this makes it tempting to always pass it implicitly in our applications.
+
+For example, here we import it at top level and all our futures will pick it up:
+
+```
+object FutureApp extends App {
+    import scala.concurrent.ExecutionContext.Implicits.global
+
+    ...
+
+    Future(blocking(1))
+
+    ...
+
+    Future(cpuBound(1))
+}
+```
+
+This may have consequences if that is not expected, for example when we want to use another executor for blocking.
+
+##### Explicitly
+
+To fix the above, we can obviously pass the blocking context explicitly and default the rest.
+
+```
+object FutureApp extends App {
+    val defaultExecutor: ExecutionContext =  scala.concurrent.ExecutionContext.Implicits.global
+    val blockingExecutor: ExecutionContext = ...
+
+    ...
+
+    Future(blocking(1))(blockingExecutor)
+
+    ...
+
+    Future(cpuBound(1))
+}
+```
+
+This is a micro management approach that can work when dealing with small code bases or teams.
+
+##### Explicitly implicitly
+
+We can improve on the above by typing our futures and thus marking them as blocking. This makes re utilization safer.
```

Comment:
I consulted the Internet and it told me reutilization doesn't have a whitespace :P please remove it.